### PR TITLE
feat: support multiple pubsub providers

### PR DESCRIPTION
**Pull request by StarpTech**

### CODE REVIEW

**Review by devsergiy — APPROVED** [review]

**Comment by devsergiy** [comment]
It looks okay for me

@jensneuse please have a look at context changes

File: v2/pkg/engine/datasource/pubsub_datasource/pubsub_datasource_test.go [hunk 1]
```diff
@@ -53,32 +49,48 @@ func TestPubSub(t *testing.T) {
...
-				Type:       EventTypeRequest,
-				TypeName:   "Query",
+				Metadata: &EventMetadata{
+					ProviderID: "default",
+					FieldName:  "helloQuery",
+					Type:       EventTypeRequest,
+					TypeName:   "Query",
+				},
+				Configuration: &NatsEventConfiguration{
+					Subjects: []string{"helloQuery.{{ args.id }}"},
```
Comment (jensneuse) [inline v2/pkg/engine/datasource/pubsub_datasource/pubsub_datasource_test.go:59]: The EDFS spec allows patterns like `helloQuery.{{ args.input.id }}` where the field argument has a root arg (input) with an input type that contains a field "id". Is this also supported by your implementation?

File: v2/pkg/engine/datasource/pubsub_datasource/pubsub_kafka.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+func (s *KafkaPublishDataSource) Load(ctx context.Context, input []byte, w io.Writer) error {
+	var publishConfiguration KafkaPublishEventConfiguration
+	err := json.Unmarshal(input, &publishConfiguration)
+	if err != nil {
+		return err
+	}
+
+	if err := s.pubSub.Publish(ctx, publishConfiguration); err != nil {
+		return err
```
Comment (jensneuse) [inline v2/pkg/engine/datasource/pubsub_datasource/pubsub_kafka.go:60]: Should this return an error or should we do this?
```go
_, err = io.WriteString(w, `{"success": false}`)
```

File: v2/pkg/engine/resolve/resolve.go [hunk 3]
```diff
@@ -422,10 +424,14 @@ func (r *Resolver) handleRemoveSubscription(id SubscriptionIdentifier) {
 		trig := r.triggers[u]
 		for ctx, s := range trig.subscriptions {
 			if s.id == id {
-				s.mux.Lock()
-				s.writer.Complete()
-				s.writer = nil
-				s.mux.Unlock()
+
+				if ctx.Context().Err() == nil {
```
Comment (jensneuse) [inline v2/pkg/engine/resolve/resolve.go:430]: You could do this check above the loop.
If we're closing 10k subscriptions, it's unnecessary.

File: v2/pkg/engine/resolve/resolve.go [hunk 4]
```diff
@@ -450,10 +456,14 @@ func (r *Resolver) handleRemoveClient(id int64) {
 	for u := range r.triggers {
 		for c, s := range r.triggers[u].subscriptions {
 			if s.id.ConnectionID == id && !s.id.internal {
-				s.mux.Lock()
-				s.writer.Complete()
-				s.writer = nil
-				s.mux.Unlock()
+
+				if c.Context().Err() == nil {
```
Comment (jensneuse) [inline v2/pkg/engine/resolve/resolve.go:462]: Check above loop?

File: v2/pkg/engine/resolve/resolve.go [hunk 5]
```diff
@@ -497,10 +507,13 @@ func (r *Resolver) shutdownTrigger(id uint64) {
 	}
 	count := len(trig.subscriptions)
 	for c, s := range trig.subscriptions {
-		s.mux.Lock()
-		s.writer.Complete()
-		s.writer = nil
-		s.mux.Unlock()
+		if c.Context().Err() == nil {
```
Comment (jensneuse) [inline v2/pkg/engine/resolve/resolve.go:522]: Check above loop?

**Review by jensneuse — CHANGES_REQUESTED** [review]

File: v2/pkg/engine/datasource/pubsub_datasource/pubsub_datasource_test.go (hunk 1 above)
Comment (StarpTech) [reply v2/pkg/engine/datasource/pubsub_datasource/pubsub_datasource_test.go:59]: Everything works as before in terms of NATS. I didn't touch this. I focused on the abstraction. Kafka won't support pattern in topics due to its different model.

File: v2/pkg/engine/datasource/pubsub_datasource/pubsub_kafka.go (hunk 2 above)
Comment (StarpTech) [reply v2/pkg/engine/datasource/pubsub_datasource/pubsub_kafka.go:60]: Good point. I like your approach. Let's return false.

**Review by jensneuse — APPROVED** [review]

**Pull request merged**